Core/include/Acts/TrackFinding/MeasurementSelector.hpp

```diff
@@ -106,60 +107,81 @@ class MeasurementSelector {
       return CombinatorialKalmanFilterError::MeasurementSelectionFailed;
     }
 
+    assert( !cuts->chi2CutOff.empty() );
+    const auto& chi2CutOff = cuts->chi2CutOff;
+    auto maxChi2Cut  = *std::max_element(chi2CutOff.begin(), chi2CutOff.end());
     double minChi2 = std::numeric_limits<double>::max();
     size_t minIndex = 0;
-    size_t index = 0;
-    // Loop over all measurements to select the compatible measurements
-    for (auto& trackState : candidates) {
-      // Take the parameter covariance
-      // const auto predicted = tackState.predicted();
-      // const auto predictedCovariance = trackState.predictedCovariance();
-
-      double chi2 = calculateChi2(
-          // This abuses an incorrectly sized vector / matrix to access the
-          // data pointer! This works (don't use the matrix as is!), but be
-          // careful!
-          trackState
-              .template calibrated<MultiTrajectoryTraits::MeasurementSizeMax>()
-              .data(),
-          trackState
-              .template calibratedCovariance<
-                  MultiTrajectoryTraits::MeasurementSizeMax>()
-              .data(),
-          trackState.predicted(), trackState.predictedCovariance(),
-          trackState.projector(), trackState.calibratedSize());
-
-      trackState.chi2() = chi2;
-
-      // Search for the measurement with the min chi2
-      if (chi2 < minChi2) {
-        minChi2 = chi2;
-        minIndex = index;
-      }
-
-      index++;
+    auto trackStateIterEnd = candidates.end();
+    {
+       auto trackStateIter = candidates.begin();
+       // Loop over all measurements to select the compatible measurements
+       // Sort track states which do not satisfy the chi2 cut to the end.
+       // When done trackStateIterEnd will point to the first element that
+       // does not satisfy the chi2 cut.
+       assert(trackStateIter != trackStateIterEnd);
+       for (;;) {
```

Comment:
I suppose this could still be
```suggestion
       for (trackStateIter = candidates.begin();trackStateIter != trackStateIterEnd;) {
```
? It's still potentially an infinite loop (if the loop body is bugged) but a bit more explicit up top?

Reply:
With future changes the loop could be turned into an infinite loop, In the current state either the front iterator or the end iterator are advanced and in each loop iteration the distance shrinks. both branches will break the loop if the distance is zero. Before the first iteration the loop cannot be empty because the entry requires that candiates is not empty. Not that it matters much wrt. the other computations but it would still add and extra comparison which was already done.
I could change it to:
```
for(  auto trackStateIter = candidates.begin(); ;) {
     assert(trackStateIter != trackStateIterEnd);
}
```
to make the intention clear.